File: Graph/DetectCycleInDirGraph/isCyclic.cpp
#include <vector>
#include <queue>
using namespace std;

class Solution {
  public:
    bool isCyclic(int V, vector<vector<int>> &edges) {
        // code here
        vector<int>indegree(V,0);
        int totalEdges = edges.size();
        vector<vector<int>>adj(V,vector<int>());
        for(int i=0;i<totalEdges;i++){
            indegree[edges[i][1]]++;
            adj[edges[i][0]].push_back(edges[i][1]);
        }
        queue<int>qu;
        for(int i=0;i<V;i++){
            if(indegree[i]==0)qu.push(i);
        }
        while(!qu.empty()){
            int ele = qu.front();
            qu.pop();
            for(int neighbour: adj[ele]){
                indegree[neighbour]--;
                if(indegree[neighbour]==0)qu.push(neighbour);
            }
        }
        for(int i=0;i<V;i++){
            if(indegree[i]!=0)return true;
        }
        return false;
    }
};

File: Graph/DetectCycleInDirGraph/isCyclic.py
from collections import deque

class Solution:
    def isCyclic(self, V, edges):
        indegree = [0] * V
        adj = [[] for _ in range(V)]

        # Build graph
        for u, v in edges:
            adj[u].append(v)
            indegree[v] += 1

        # Push all vertices with indegree 0
        q = deque()
        for i in range(V):
            if indegree[i] == 0:
                q.append(i)

        # Kahn's Algorithm
        while q:
            node = q.popleft()

            for neighbour in adj[node]:
                indegree[neighbour] -= 1
                if indegree[neighbour] == 0:
                    q.append(neighbour)

        # If any vertex still has indegree > 0, cycle exists
        for i in range(V):
            if indegree[i] != 0:
                return True

        return False